activation/handler_v1.go

```diff
@@ -330,90 +330,243 @@ func (h *HandlerV1) cacheAtx(ctx context.Context, atx *types.ActivationTx) *atxs
-					},
-					SmesherID: atx.SmesherID,
-					Signature: prevSignature,
-				}, {
-					InnerMsg: types.ATXMetadata{
-						PublishEpoch: atx.PublishEpoch,
-						MsgHash:      atx.ID().Hash32(),
-					},
-					SmesherID: atx.SmesherID,
-					Signature: signature,
-				}},
+			atx2, err := atxs.Get(tx, id)
+			if err != nil {
+				return nil, fmt.Errorf("get prev atx: %w", err)
 			}
-			proof = &mwire.MalfeasanceProof{
-				Layer: atx.PublishEpoch.FirstLayer(),
-				Proof: mwire.Proof{
-					Type: mwire.MultipleATXs,
-					Data: &atxProof,
-				},
+			if atx.ID() != atx2.ID() && atx.PrevATXID == atx2.PrevATXID {
+				// found an ATX that points to the same previous ATX
+				atx2ID = id
+				break
 			}
-			encoded, err := codec.Encode(proof)
+			pubEpoch = atx2.PublishEpoch
+		}
+	}
+
+	if atx2ID == types.EmptyATXID {
+		// something went wrong, we couldn't find an ATX that points to the same previous ATX
+		// this should never happen since we are checking in other places that all ATXs from the same node
+		// form a chain
+		return nil, errors.New("failed double previous check: could not find an ATX with same previous ATX")
+	}
+
+	var blob sql.Blob
+	v, err := atxs.LoadBlob(ctx, tx, atx2ID.Bytes(), &blob)
+	if err != nil {
+		return nil, err
+	}
+	if v != types.AtxV1 {
+		return nil, fmt.Errorf("previous atx %s is not of version 1", atx.PrevATXID)
+	}
+
+	var watx2 wire.ActivationTxV1
+	if err := codec.Decode(blob.Bytes, &watx2); err != nil {
+		return nil, fmt.Errorf("decoding previous atx: %w", err)
+	}
+
+	proof := &mwire.MalfeasanceProof{
+		Layer: atx.PublishEpoch.FirstLayer(),
+		Proof: mwire.Proof{
+			Type: mwire.InvalidPrevATX,
+			Data: &mwire.InvalidPrevATXProof{
+				Atx1: *atx,
+				Atx2: watx2,
+			},
+		},
+	}
+
+	if err := identities.SetMalicious(tx, atx.SmesherID, codec.MustEncode(proof), time.Now()); err != nil {
+		return nil, fmt.Errorf("add malfeasance proof: %w", err)
+	}
+
+	h.log.WithContext(ctx).With().Warning("smesher referenced the wrong previous in published ATX",
+		log.Stringer("smesher", atx.SmesherID),
+		log.ShortStringer("expected", prevID),
+		log.ShortStringer("actual", atx.PrevATXID),
+	)
+	return proof, nil
+}
+
+func (h *HandlerV1) checkMalicious(
+	ctx context.Context,
+	tx *sql.Tx,
+	atx *wire.ActivationTxV1,
+) (*mwire.MalfeasanceProof, error) {
+	malicious, err := identities.IsMalicious(tx, atx.SmesherID)
+	if err != nil {
+		return nil, fmt.Errorf("checking if node is malicious: %w", err)
+	}
+	if malicious {
+		return nil, nil
+	}
+	proof, err := h.checkDoublePublish(ctx, tx, atx)
+	if proof != nil || err != nil {
+		return proof, err
+	}
+	return h.checkWrongPrevAtx(ctx, tx, atx)
+}
+
+// storeAtx stores an ATX and notifies subscribers of the ATXID.
+func (h *HandlerV1) storeAtx(
+	ctx context.Context,
+	watx *wire.ActivationTxV1,
+	blob []byte,
+	received time.Time,
+	leaves uint64,
+	effectiveNumUnits uint32,
+) (*types.ActivationTx, *mwire.MalfeasanceProof, error) {
```

Comment:
How about instead:
```suggestion
func (h *HandlerV1) storeAtx(
	ctx context.Context,
	atx *types.ActivationTx, 
	watx *wire.ActivationTxV1,
	blob []byte,
) (*mwire.MalfeasanceProof, error) {
```

To avoid querying for additional data (`baseTickHeight`) and constructing the ATX here.